bookstore/archive.py

```diff
@@ -29,6 +41,46 @@ def __init__(self, *args, **kwargs):
             s3_additional_kwargs={},
         )
 
+        self._thread_pool = ThreadPoolExecutor(max_workers=30)
+
+        self.path_locks: Dict[str, Lock] = {}
+
+    async def archive(self, record: ArchiveRecord):
+        """Process a storage write, only allowing one write at a time to each path
+        """
+        lock = self.path_locks.get(record.filepath)
```

Comment:
I think this still has the potential for the race that Charles was worried about because two routines might want to create a new lock for the same path. To close that hole we'd need a lock, initialized when this object is made, that protects `self.path_locks` while we add a new path in there.

Reply:
I definitely agree, haven't gotten back to this one yet.